### ISSUE TITLE
Installing plugins from bundled app

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#1074]
## 🚀 Feature
As discussed in the last meeting, we'll need a way to extend bundled apps with plugins.  So this issue is here to propose and discuss ideas.

After playing with it a little bit, I have one possible solution:
As long as `pip` is included in the bundle (easily done by adding it to `hiddenimports` in the pyinstaller spec), people can pip install packages... but will _not_ of course, be able to install them to the frozen bundle itself (actually, I found that the default behavior was to install into the base python environment 🤮).  Pip provides both the `--target` and `--prefix` options that let you specify where things should be installed (the difference is that `--target` seems to install _all_ of the dependencies to that target directory, whereas `--prefix` first checks the existing PYTHONPATH and only installs what is missing to the prefix directory).
As a proof of principle, I just tried pip installing `napari-dv` to `~/.config/napari/plugins` and then used `napari.plugins.plugin_manager.autodiscover()`, and provided that path... and it worked (I could read `.dv` files from the bundled app).

I would propose that we have a function (& GUI) within napari that lets users install a pip or conda dependency (using `--prefix`), use [appdirs](https://github.com/ActiveState/appdirs) to get the appropriate platform-specific `user_data_dir`, then run `plugin_manager.autodiscover(path=user_data_dir)` on napari launch (or after plugin install, it wouldn't even require a relaunch after a plugin was installed 🎉).  We could also detect whether napari is being run from a bundle or not, and have the function just `pip install` into the regular current environment if not frozen.

thoughts?

(as a sidenote, i think `appdirs` would be useful for #834 as well... and we should just vendor it, it's a single file that hasn't had a pypi release in too long).

**Comment by tlambert03** [comment]
some more info here.  I was able to make a nice "`install_plugin`" function that installs packages with pip to some platform-specific folder (outside of the environment), which then loads fine with our `plugin_manager.discover()` method.  
However, it's going to be rather tricky to "bundle" pip with pyinstaller for a couple reasons:
- pip _strongly_ discourages using anything but the CLI to install things... they don't support programmatic install with anything other than `subprocess.run(['python', '-m', 'pip', 'install'])`.  (Nonetheless, I tried bundling `pip`, but it didn't work... just hung when I tried to install).
- however, I don't yet know how to get at "`python`" within a bundled app.  normally, you could use `sys.executable` to access the current interpreter, but in our bundled app, `sys.executable` does not resolve to "python" but rather to `napari.__main__`
- EDIT: also... simply using `subprocess.run(['python', '-m', 'pip', 'install'])` works, but will use whatever the system python is, and if it differs from the version that was used to bundle the app, we have issues.

This is not a deal breaker, as we can simply "roll our own" pip-like thing for bundled apps.  All that needs to be done for wheels is to download them from pypi and extract them to our plugins folder, then it all works.  Uninstallation is as simple as deleting the extracted files.  I haven't yet figured out what non-wheel installations would take, but it seems like a reasonable starting point to only support wheel plugins in our bundled app.

**Comment by jni** [comment]
@tlambert03 

> however, I don't yet know how to get at "python" within a bundled app. 

Does `sys.base_exec_prefix` point to something useful in this context?

**Comment by jni** [comment]
By the way, though, using pip might be super dangerous in this context. If a plugin declares a dependency that conflicts with our existing stack, that dependency will get overwritten, I think? Fun! It would be nice to have a way to tell pip, don't touch existing packages, but rather, error on install.

**Comment by tlambert03** [comment]
> Does sys.base_exec_prefix point to something useful in this context?

doesn't seem to.  on a mac, if the bundled app is at `./napari.app`, then `sys.base_exec_prefix` points to `./napari.app/Contents/MacOS` and `sys.executable` points to `./napari.app/Contents/MacOS/napari`.  

> By the way, though, using pip might be super dangerous in this context. If a plugin declares a dependency that conflicts with our existing stack, that dependency will get overwritten, I think? Fun! It would be nice to have a way to tell pip, don't touch existing packages, but rather, error on install.

as long as `--prefix` is used (which is what I was doing with the bundled app) then any new packages seemed to be installed only into `prefix`... but I'll keep playing with that.

**Comment by tlambert03** [comment]
just want to document this somewhere:
I've been playing around with bundling `pip` in the pyinstaller bundle, with the hopes that we will be able to offer a "install plugin" feature in the bundled app, which would install to some external environment (probably in [`appdirs.user_data_dir`](https://github.com/ActiveState/appdirs)).  The good news is that I can successfully bundle the full pip executable (since [pip must be run as a subprocess](https://pip.pypa.io/en/latest/user_guide/#using-pip-from-your-program)) and install packages to some external directory, _and_ the bundled app can discover and use plugins installed there.  So that's great.

The problem at the moment is that the bundled `pip` is not aware of all of packages installed in the bundle.  So when you `pip install some-plugin`, if that plugin depends on napari (or numpy or something), pip thinks it needs to install _everything_, and you get napari etc installed in the plugin directory.  I dug through the source on how pip discovers installed packages, and found that ultimately it depends on [`pkgutil.get_importer`](https://github.com/pypa/pip/blob/69cc55831d529aaec13e0fb42c763ed103edd457/src/pip/_vendor/pkg_resources/__init__.py#L1967-L1971).    this issue https://github.com/pyinstaller/pyinstaller/issues/1905 nicely explains why pip isn't discovering anything:  the frozen `pkgutil.iter_modules` is only supported for Python's built in loaders.  It provides some workarounds for detecting bundled packages, but it's still unclear to me how we would hand those to the `pip` CLI.

**Comment by tlambert03** [comment]
another challenge with pip installing additional packages in a frozen environment:
the whole point of pyinstaller is to bundle the minimum number of modules and libraries required to run the `__main__.py` script (and if one wants to include additional stuff, they must explicitly include it with [`hiddenimports`](https://pythonhosted.org/PyInstaller/when-things-go-wrong.html#listing-hidden-imports).
This means that even if we tell a "frozen" version of pip which packages were included in the bundle (meaning it doesn't need to reinstall them), some of those packages may not actually be "feature complete" if the hook for the corresponding package misses some hidden imports.  
Probably not a huge problem, but could mean that we're often chasing dependency bugs for plugins in a bundled environment.  I don't immediately see a way around this, since the whole concept of a site-packages folder with distributions doesn't really exist in a frozen environment.

**Comment by tlambert03** [comment]
@sofroniewn, I think i could probably use some input from your connections on the pip team at this point.  Here's how I would summarize the problem:

- when we create a bundled app with pyinstaller, it [analyzes our main script](https://pyinstaller.readthedocs.io/en/stable/operating-mode.html#what-pyinstaller-does-and-how-it-does-it) and bundles only those modules and libraries required to run it.
- _after_ bundling, we'd like users of the bundled app to be able to `pip install` new (plugin) packages.  (Assume we bundle pip with our program and use the `pip install --prefix` option to put things in some directory external to the bundle which we will then include in `sys.path`).
- **question:** how do we tell the bundled `pip` which of the packages in the dependency chain it doesn't need to reinstall because they already exist in the bundle?  (Running `pip list` within the bundle shows _only_ `pip` being installed.  Presumably this is because all of the `pkg_resources` and `pkgutil` functions [don't work the same way in a bundled app](https://github.com/pyinstaller/pyinstaller/issues/1905)... is there a possible workaround, or is this a dead end?)

**Comment by tlambert03** [comment]
I've been working on an alternate strategy to pyinstaller: I have a working (mac) prototype ([download here](https://api.cirrus-ci.com/v1/artifact/task/5474815058378752/napari/dist/napari.dmg)) and I'm curious to get people's thoughts.  

I stumbled onto [this blog post](https://dschreij.github.io/how-to/package-anaconda-environments-as-apps), that describes a _non_-frozen (mac-specific for now) approach to app bundling, which essentially boils down to just putting a complete conda environment into the `bundle.app/Contents/Resources` folder.  It's definitely unconventional, and will be unlikely to get into the AppStore (so will always have that scary security warning on launch), but has some advantages:

1) tremendous flexibility in the post-bundled environment.  It's literally just a conda environment with all the original source code.  You can pip install, or pip upgrade stuff as usual...
2) very predictable behavior: in everything I've tried so far, it behaves just like it does in a conda env (which definitely can't be said for pyinstaller... which has required tons of trial and error)
3) it's very easy to debug... since you can just edit the code in the bundle to test stuff out (whereas pyinstaller bundles pre-compiled code).

downsides of course include:
1) bigger bundle size, maybe 2x as big?  (napari is ~1GB uncompressed)
2) unconventional approach, will likely make app stores unhappy... but probably shouldn't be different than Fiji.app

the original blog had some code that worked for a specific project, but I generalized it into a package that aims to create a mac bundle for any pip installable package (it also works great [on CI](https://github.com/tlambert03/conda-bundler/blob/master/.cirrus.yml) with really minimal setup): https://github.com/tlambert03/conda-bundler

If you have a mac, [download the app](https://api.cirrus-ci.com/v1/artifact/task/5474815058378752/napari/dist/napari.dmg), and give it a try.  You can pip install anything within the app console using
```python
import subprocess, sys
subprocess.run([sys.executable, '-m', 'pip', 'install', 'napari-dv'])

# and if you want to register that new plugin without reboot
from napari.plugins import plugin_manager
plugin_manager.discover()
```

**Comment by jni** [comment]
@tlambert03 that is so, so cool. I'm practically in tears. 😂 

I *do* want a 1GB executable containing all of Anaconda for napari. That will make our dependency messaging so much easier. Again, we need to be super careful that we don't clobber our existing installs, e.g. if we ship NumPy 1.17 and some plugin declares 1.18, **installation should fail**, rather than clobber our shipped NumPy. Annoyingly I don't see a way to tell pip "never upgrade dependencies", *but*, I do see this option:

```
  -c, --constraint <file>     Constrain versions using the given constraints
                              file. This option can be used multiple times.
```

So perhaps we can build a constraint file containing all of Anaconda's shipped versions, and that will solve the problem.

Now we (ie you? =P) need to pull off the same thing for Windows, and I will be so happy. =)

Finally, re the scary launch message, I *think* becoming an "identified developer" helps with that. Perhaps @ctrueden or @0x00b1 have advice on this?

**Comment by ctrueden** [comment]
> unlikely to get into the AppStore (so will always have that scary security warning on launch)
...
> I think becoming an "identified developer" helps with that.

You don't have to distribute via the app store, but you do need to code-sign the app with your Developer ID Certificate, which you get by joining Apple's developer program for $99/year.

https://developer.apple.com/developer-id/

For Fiji, I have resisted doing this out of principle, meaning new Fiji users on macOS always have to:
1. Copy the app into Applications.
2. Right-click the app and choose Open.
3. When the popup appears complaining about the app being from an unidentified developer, choose Open.

On Catalina, some people have reported needing to perform [additional steps](https://forum.image.sc/t/warning-your-imagej-installation-cannot-be-updated-because-it-is-read-only/20333/4) to remove the app from initial quarantine.

So you can either add these instructions to your installation documentation, or you can fork over the $99/year and sign the app.

Finally, there is the [issue of app notarization on Catalina](https://appleinsider.com/articles/19/12/23/apple-will-enforce-app-notarization-for-macos-catalina-in-february), which may be creating additional difficulties as of February. I haven't had time to investigate it yet. But fortunately if [this guide for Electron apps](https://www.update.rocks/blog/osx-signing-with-travis/) is any indication, it looks like it's possible to both code-sign and notarize apps automatically via CI.

@0x00b1 How has CellProfiler tackled these issues?

**Comment by GenevieveBuckley** [comment]
I gave your bundled app a go on my Mac @tlambert03, here's what happened:

1. I downloaded the dmg file you linked to at https://api.cirrus-ci.com/v1/artifact/task/5474815058378752/napari/dist/napari.dmg
2. I double clicked it to run the installer, everything looks fine & I can drag the bundled napari app into my Applications folder
3. I start napari either from the launchpad, or by clicking it in the Applications folder. The first time, I get a "Veryifying napari.app" message & then an error message that the app can't be opened because the developer could not be identified. The second time you try to launch napari, it goes straight to the error message.

<img width="405" alt="2020-04-03_napari_verifying_developer" src="https://user-images.githubusercontent.com/30920819/78320890-4fb54980-75b6-11ea-9df0-9e5d9cd83659.png">

<img width="426" alt="2020-04-03_napari_bundle_error" src="https://user-images.githubusercontent.com/30920819/78320896-5348d080-75b6-11ea-8f9c-7b1749964ce4.png">


Right clicking napari in the Applications folder shows me the usual menu you get when your right click a thing, it doesn't have anything that looks useful here.

I'm no stranger to installing random things that did not come from the app store, but this isn't like any of those. I don't think your digital signature worked properly? Let me know if/when you want me to try again, I'm happy to help :)

**Comment by jni** [comment]
There's no digital signature in this bundle. Apple just keeps changing the rules for unsigned apps. Now you have to right-click -> Open, and then you should get a dialog that takes you to system preferences to add an exception for this app. It is super annoying.

**Comment by GenevieveBuckley** [comment]
Yes, that's what I'm used to - right-clicking to get the option to run the app anyway.

What actually happens is:
1. I right-click the napari.app in the Applications folder, and select "Open"
2. The "verifying napari.app" message appears, if this is the first time
3. The error message appears: '"napari.app" cannot be opened because the developer could not be identified'.

(I've just re-downloaded & redone this from the beginning, in case I'd made some kind of silly mistake, but this is happening the same way every time.

**Comment by jni** [comment]
Maybe now they've been even more helpful and you have to go to system preferences > security manually. Or maybe you have to hold down option.

**Comment by GenevieveBuckley** [comment]
Oh wow, they have changed the way you can do this! Yes, system preferences > security and privacy > general. *Then* you can use your administrator password to unlock the settings and click the "Run Anyway" button. Whew!

A few years ago it was: right click > Open, select "Open Anyway" in the popup dialog.
Even more recently it was just double click, select "Open Anyway" in the popup dialog.

**Comment by tlambert03** [comment]
Just to clarify, this app IS digitally signed... but it’s done using an ad hoc signature.  (See [this post](https://eclecticlight.co/2019/01/17/code-signing-for-the-concerned-3-signing-an-app/) for a good post explaining the three levels of signing).   You actually *have* to use some level of signing for it to open at all... took me a while to figure that out.
I also tried using a self-signed certificate, which is a bit better than ad hoc (but I didn’t take the time to do that on CI).  Ultimately, we would like to sign with a certificate signed by Apple, which I _think_ will make this app behave a bit more like you’re used to Genevieve, where you at least get the “open anyway” notice without having to manually go into System prefs.

But what you observed here is how it looks on my computer too.  Did it eventually run for you?

**Comment by GenevieveBuckley** [comment]
It did run! 

I didn't do very much with it, just opened an image and mucked around. Let me know if you have specific things you want checked (like the pims bioformats stuff later on).

**Comment by tlambert03** [comment]
OK, promising news for windows!  I was able to create a windows installer with [inno setup](https://jrsoftware.org/isinfo.php) that bundles the conda environment similarly to the mac strategy described above.  I installed it on a different windows computer and it all seems to work! ([download installer here](https://www.dropbox.com/s/jqqm2tbahh790q3/napari_installer.exe?dl=0), if you want to try)

the main difference between mac and windows is that, while mac encourages apps to use their app folder for file storage, windows prohibits it without running the app with admin rights.  That's not a huge problem, since we can fall back to the `appdirs` & `plugin_manager.autodiscover(path=user_data_dir)` strategy that I described in the first post in this thread.  But it raises a question of consistency.  Since we _have_ to do that on windows, and we _can_ do it on mac if we want to, perhaps we should just make that the default place that pip will install any additional packages to (this would be like `~/Library/App Support` on mac), and add it to sys.path on launch.  Anyway, sorta minor details.

there are a couple of "niceties" that I still need to fix for the windows app, like the icon doesn't work, and i don't have an automated installed that works on CI, but i think it's going to work.

**Comment by tlambert03** [comment]
just to mention it again here: my current hope for linux is that we can use [snapcraft.io](https://snapcraft.io/) to bundle napari (and at one point I had a [working prototype](https://snapcraft.io/napari) for that), but I still haven't experimented with how that handles pip installs after the user installs napari.

**Comment by GenevieveBuckley** [comment]
Re: windows app
It takes a solid 17 seconds to even display the splash logo when I try and launch napari (the actual viewer appears about a second after that). That's a big deal in terms of the user experience, I genuinely thought the app had failed to launch at all. (Windows 10 laptop, older but still decent, i7 CPU, 16GB RAM)

I don't think you should make the Mac apps install plugins to a location outside their app bundle. Just because Windows forces you to do an ugly thing isn't a good enough reason to do the same ugly thing everywhere, imo. If Linux also has the same restriction, then maybe it's enough to consider it, I dunno.

**Comment by tlambert03** [comment]
Did you try it a second time?  I believe that’s all the python source compiling the first time.  17 seconds is definitely longer than what I observed, but yeah it’s slow the first time, but then only about a second for the splash screen after for me... you?

I’m not really sure what to do about that

**Comment by tlambert03** [comment]
I’m also not entirely convinced that installing plugins to some external folder is doing the ugly thing.  It does mean that the bundle remains “pure” and it’s very easy to return it to the default state (i.e., user installs remain physically separated from bundled packages, instead of everything in the same site-packages folder).  Also, if they delete the bundle and reinstall, I believe they wouldn’t need to reinstall other stuff (provided it’s the same python version in the new bundle) with the external folder approach

**Comment by GenevieveBuckley** [comment]
> 
> Did you try it a second time? I believe that’s all the python source compiling the first time. 17 seconds is definitely longer than what I observed, but yeah it’s slow the first time, but then only about a second for the splash screen after for me... you?
> 
> I’m not really sure what to do about that

That was the second or third time. Timed it again a few times just now with a stopwatch & it's around 8 - 10 seconds. No idea why that's so different than earlier (I guess I could be running less other stuff?)

**Comment by GenevieveBuckley** [comment]
> 
> 
> I’m also not entirely convinced that installing plugins to some external folder is doing the ugly thing. It does mean that the bundle remains “pure” and it’s very easy to return it to the default state (i.e., user installs remain physically separated from bundled packages, instead of everything in the same site-packages folder). Also, if they delete the bundle and reinstall, I believe they wouldn’t need to reinstall other stuff (provided it’s the same python version in the new bundle) with the external folder approach

Hm, persistence of plugins (and other settings?) could be really useful. That's definitely an argument in favour.

This discussion might be a good one for the next group meeting, I don't want to side track you here.

**Comment by tlambert03** [comment]
> That was the second or third time. Timed it again a few times just now with a stopwatch & it's around 8 - 10 seconds. No idea why that's so different than earlier (I guess I could be running less other stuff?)

Omg, that’s no good, and also very strange.  Not sure what could be causing that, since at that point it should be behaving exactly like a conda env (the “exe” is just a bat file that calls “python napari”).  

The problem here is that even if we wanted to throw up a loading indicator, I’m not sure how to get it up any earlier than that splash screen.  So well need to figure out what’s causing that delay somehow.

**Comment by GenevieveBuckley** [comment]
Yeah, this is tricky.

The majority of the time is being taken up by that "python napari" call. The outer layers wrapped around it (the .bat file, and the 'runhidden.vbs') don't add much extra time. I didn't dig any further than that.

**Comment by GenevieveBuckley** [comment]
Also, does BeeWare's briefcase look useful for your purposes? https://github.com/beeware/briefcase

Unlike pyinstaller, things aren't frozen. There's more discussion of that here: https://github.com/beeware/briefcase/issues/145

**Comment by tlambert03** [comment]
that _does_ look interesting!  it always amazes me how you can look into something like this for days and still miss a 700+ star project that seems to do exactly what you're looking for.  I'll take a closer look, thanks for mentioning it!

**Comment by GenevieveBuckley** [comment]
> it always amazes me how you can look into something like this for days and still miss a 700+ star project that seems to do exactly what you're looking for. I'll take a closer look, thanks for mentioning it!

That happens to me all the time, too!

I found out about the project when I met Russell at PyConAU last year. They have a gitter, we can probably ask questions there if we need.

So far all I've done is go through the tutorial, and run briefcase on napari for Windows & Mac (basically all you have to do is change `pyproject.toml`). I haven't done anything to look into how installable plugins would work, and if that's the same or different as your conda bundles. I'm interested to know how you think it compares.

**Comment by tlambert03** [comment]
the main way I've been testing plugins is as described in https://github.com/napari/napari/issues/1074#issuecomment-606162357

use the following code in the bundled built-in napari console, and see if and where the package gets installed

```python
import subprocess, sys
subprocess.run([sys.executable, '-m', 'pip', 'install', 'napari-dv'], capture_output=True)
```

**Comment by GenevieveBuckley** [comment]
I've been tinkering around more with briefcase, and it deliberately doesn't package pip in with python in the applications.

It might *technically* be possible to fork & tinker with the python support packages for Apple/Android/Linux/Windows, but this doesn't seem like less work than continuing with your separate approaches for different platforms.

FYI: This [comment](https://github.com/beeware/briefcase/issues/383#issuecomment-619780148) on [issue #383](https://github.com/beeware/briefcase/issues/383) outlines the project's position on this:
> The macOS (and iOS, and Android) builds that Briefcase uses are all managed by BeeWare; we've mostly followed the lead of the Windows installer in terms of omissions. In fact, having pip in the iOS binary could actually constitute grounds for ejection from the iOS App Store, as they prohibit using any mechanism to get around App Store code review.
> ...
> pip and venv are another matter; there's definitely a subset of use cases (and, admittedly, Mu is one of them) for which including pip and venv make sense; but for the default "user application" case, they're "overhead with security implications".

**Comment by sofroniewn** [comment]
@0x00b1 I just heard from @neuromusic that you might be available to help napari with our bundled app, which would be very exciting!

One thing that we're really keen on is allowing people to pip install into the bundle after download. For this reason we're no longer looking at pyinstaller which makes a "frozen app". Instead @tlambert03 did some amazing work to make a proof of concept described here https://github.com/napari/napari/issues/1074#issuecomment-606162357 for "non-frozen app" for mac. @GenevieveBuckley subsequently found [briefcase](https://github.com/beeware/briefcase) mentioned here https://github.com/napari/napari/issues/1074#issuecomment-610931682 which would do some of this process for us and we might want to leverage.

I'd love to get your take on these approaches and it would be great if you wanted to dive in and help us get this across the finish line, particularly for windows (see https://github.com/napari/napari/issues/1074#issuecomment-610098978 for some initial progress from @tlambert03) and linux (see https://github.com/napari/napari/issues/1074#issuecomment-610099849 from @tlambert03 too). @tlambert03 do you have anything else to add?

**Comment by GenevieveBuckley** [comment]
I'll give briefcase another go, following on from the discussion at our meeting today.

**Comment by GenevieveBuckley** [comment]
I've had success using briefcase to create a napari Mac app that can install & run new plugins. Hooray!

### The napari Mac app on first launch

<img width="1095" alt="01-napari-bundle-default-plugins-only" src="https://user-images.githubusercontent.com/30920819/82744431-e4207900-9dbb-11ea-859b-2c4a0531b568.png">

### After installing a new plugin

<img width="1088" alt="04-napari-bundle-after-plugin-installed" src="https://user-images.githubusercontent.com/30920819/82744444-031f0b00-9dbc-11ea-9867-2b2ac200f938.png">

### Checking the napari-dv plugin works

<img width="1096" alt="05-napari-bundle-working-plugin" src="https://user-images.githubusercontent.com/30920819/82744452-221d9d00-9dbc-11ea-890d-95b7da172e92.png">

**Comment by sofroniewn** [comment]
That's amazing @GenevieveBuckley! I'm very excited about this!!

**Comment by tlambert03** [comment]
Hooray!  I tried last night as well and worked for me too on Mac, just had to add pip to the requirements in toml right?  From your reading do you know where we should expect that whole pip thing to pop up for us?  Is it a windows specific thing?  I also successfully built with pip in the bundle on Windows, but had an as of yet unidentified problem when trying to run the post installed app

**Comment by GenevieveBuckley** [comment]
> just had to add pip to the requirements in toml right? 

Yes, that's right. 
FYI: I'm really happy about that, I was worried we'd have to modify the python support repositories for every platform individually. So this was much easier than I expected!

> From your reading do you know where we should expect that whole pip thing to pop up for us? Is it a windows specific thing? 

I'm sorry, what do you mean by "that whole pip thing to pop up for us"?

> I also successfully built with pip in the bundle on Windows, but had an as of yet unidentified problem when trying to run the post installed app

I've also built & packaged it successfully on Windows (there were a handful of font file warnings in the packaging). BUT when it comes time to run napari from the packaged app, I can't open the integrated ipython console 😢 . So that's frustrating, but I'm working on it.

Is that the same as your unspecified problems?

**Comment by tlambert03** [comment]
> I'm sorry, what do you mean by "that whole pip thing to pop up for us"?

never mind, I was just wondering where the concern in https://github.com/napari/napari/issues/1074#issuecomment-620968132 came from... but i think your last post answered it.

my windows issue is that the app didn't launch at all, but I doing it over chrome remote desktop, so there were confounds... 

I'm gonna start working on having this run on CI, and add the artifacts to your releases, if you come across good tricks with the .toml file, we can merge our findings into a PR.

**Comment by GenevieveBuckley** [comment]
> I'm gonna start working on having this run on CI, and add the artifacts to your releases, if you come across good tricks with the .toml file, we can merge our findings into a PR.

Yes, definitely! The branch I have is [here](https://github.com/GenevieveBuckley/napari/tree/briefcase-packaging), if you want to work off it or merge stuff into it.

**Comment by tlambert03** [comment]
cool.  I see you've also hard-coded the version string into the toml... that's something we'll need to figure out.  thoughts?

**Comment by GenevieveBuckley** [comment]
I think I've done a fairly comprehensive investigation now. Here's the state of things so far:

Desktop platforms
* Mac app - works great! See [this comment](https://github.com/napari/napari/issues/1074#issuecomment-633168979) for some gorgeous screenshots.
    * 236.2 MB  .dmg package installer size
    * 940.5 MB in Application folder
    * New plugins are installed into the `site-packages` folder for python in  the bundled app
* Windows app - sort of works? I can package and launch, but am unable to open the integrated ipython console from the packaged app. It's also flicker-y and weird on launching, a terminal seems to flash up on the screen about four times before you finally see the napari splash logo and napari starts. So that's not ideal.
    * 240 MB .msi package installer size
    * 641 MB Windows app installed into C:\Users\USERNAME\AppData\Local\Programs\napari
    * Can @sofroniewn suggest anything we should check re: the ipython console?
* Linux app - so close, but no build yet!
    * I've been using VirtualBox to create an Ubuntu VM, but am missing libgdbm3.so & can't work out how to install it. Shouldn't this be a simple
    * Can someone running linux give this a go? It's super simple, just 6 bash commands and then  checking if you can launch the packaged app. I'll put a comment below.

Mobile platforms
* iOS - nope, briefcase assumes that the app is an executable module which creates the ObjectiveC PythonAppDelegate object. That assumption does not match what PySide/PyQt is actually doing ([more details here](https://github.com/GenevieveBuckley/hellopyside/issues/1)).
* Android - nope, briefcase currently has spotty support for Android & I wasn't able to get a hello world app to run in the simulator.

One alternative option for mobile could be [pyqtdeploy](https://www.riverbankcomputing.com/software/pyqtdeploy/). The downside is it freezes everything during packaging, so no installable plugins. Mind you, iOS is very strict about this as well, installable plugins may never have been an option there anyway.

**Comment by GenevieveBuckley** [comment]
# Call for volunteers! 
## Can someone running linux test our app packaging?

It's pretty straighforward, just make an empty directory and type this from the command line:
```
pip install briefcase
git clone --single-branch --branch briefcase-packaging https://github.com/GenevieveBuckley/napari.git
cd napari
briefcase dev  # check briefcase can run napari before it gets packaged
briefcase create
briefcase build
briefcase run  # runs the packaged napari app 
```

Last, we want to test if we can install new plugins into the packaged app:
1. Install the napari package like you would install any other new software to your system
2. Launch napari and screenshot the list of all installed plugins (from the napari Plugins menu).
3. Open the integrated ipython console, and type:
```python
import subprocess, sys
subprocess.run([sys.executable, '-m', 'pip', 'install', 'napari-dv'], capture_output=True)
```
4. Screenshot the list of installed plugins again, this time "dv_reader" should be added to list.

[Here's an example file](https://github.com/tlambert03/napari-dv/blob/master/toxo.dv) you can drag and drop into napari after installing the napari-dv plugin.

**Comment by GenevieveBuckley** [comment]
@tlambert03 you should also know there is a [pytest-briefcase](https://github.com/beeware/pytest-briefcase) plugin, this is probably something we want included in our packaging & bundle release pipeline.

It's so new it's not even on PyPI yet (but will be after the next release of briefcase is available).

**Comment by tlambert03** [comment]
yeah, this one is confusing.  libgdbm3 is for ubuntu 16.04, and 18.04 uses libgdbm5 (and I actually can't seem to force install the older version)

the linux packaging is actually done by https://github.com/linuxdeploy/linuxdeploy.  They seem to test (CI) on 16.04 ... though there are certainly issues suggesting it should work on 18.04

**Comment by tlambert03** [comment]
I can confirm that linux builds correctly on github actions with ubuntu 16.04 (but will fail otherwise), the key is in the beeware tutorial under the linux tab [here](https://docs.beeware.org/en/latest/tutorial/tutorial-3.html#creating-your-application-scaffold).

I was able to get all three platforms to build successfully on github actions with [downloadable artifacts](https://github.com/tlambert03/napari/actions/runs/114134108). Note, I had to actually [patch briefcase](https://github.com/tlambert03/napari/blob/wip/briefcase/patch_briefcase.py) on mac because `dmgbuild.build_dmg` doesn't seem to be creating a large enough dmg (I see this locally too, but have to edit the source code of briefcase directly to fix it... I created [an issue at briefcase](https://github.com/beeware/briefcase/issues/406)... but the patch works fine as well if a fix takes a while)


**macOS**: downloaded artifact runs fine on mac, with the major caveat that since this bundle is not using _any_ code signing (not even ad hoc signing), the only way to get it past gatekeeper is to literally run `open /Applications/napari.app/Contents/MacOS/napari` in a terminal ... (but thereafter you can just double click on the app icon).  So figuring out our code signing strategy is definitely going to be important there.

**Windows**:  the windows bundle has two issues:

1. it's not able to build icons because, frustratingly, pyside2-rcc binary didn't make it into the bundle, (in fact, it looks like _none_ of the typical setup.py `console_scripts` have made it in?).  So, @GenevieveBuckley, that "unidentified" error I was mentioning before was just that the program never launches, and nothing is shown. By calling the program directly in the windows shell `C:\Users\<username>\AppData\Local\Programs\napari\python\python.exe -m napari`, I could see the error message.  And after manually putting in the resources file, it launched ok... so, that's something to work on...  (@GenevieveBuckley, I'm wondering why it even launched for you?  maybe you have one of the pyrcc binaries in your global environment?)
2. The console problem that @GenevieveBuckley was seeing in https://github.com/napari/napari/issues/1074#issuecomment-633231851 is because the default  "pythonw.exe" executable in briefcase is launching napari in a "hidden console" mode such that napari _thinks_ there is a console somewhere else (though it's not visible), and so the built in console is disabled (just like would it be if you launched `napari` from within an ipython session).  If you go into the properties of the shortcut icon that launches napari and change the target from `"pythonw.exe -m napari"` to `"python.exe -m napari"` (note the missing "w" there)... then the built in console works, but then it _also_ pops up an empty windows cmd shell 😖

**Linux** ... haven't tried the linux artifact yet... will update when I do
EDIT:  no joy yet for linux.  To run the AppImage, I had to first make it executable with `chmod u+x <AppImageFile>`... but after I did that, nothing happened when I double clicked it.  By running the AppImage directly in the terminal, I got this error:
```
Traceback (most recent call last):
  File "/tmp/.mount_napari6jrUf5/usr/lib/python3.8/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/tmp/.mount_napari6jrUf5/usr/lib/python3.8/runpy.py", line 142, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/tmp/.mount_napari6jrUf5/usr/lib/python3.8/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/tmp/.mount_napari6jrUf5/usr/app/napari/__init__.py", line 40, in <module>
    from vispy import app
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/vispy/__init__.py", line 30, in <module>
    from .util import config, set_log_level, keys, sys_info  # noqa
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/vispy/util/__init__.py", line 9, in <module>
    from .logs import logger, set_log_level, use_log_level  # noqa
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/vispy/util/logs.py", line 14, in <module>
    import numpy as np
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/core/__init__.py", line 102, in <module>
    from . import _dtype_ctypes
  File "/tmp/.mount_napari6jrUf5/usr/app_packages/numpy/core/_dtype_ctypes.py", line 25, in <module>
    import _ctypes
ModuleNotFoundError: No module named '_ctypes'
```
don't yet know what's going on there... maybe something to do with [`libffi`](https://stackoverflow.com/questions/27022373)?  Or maybe something to do with having built on 16.04 and trying to run on 18.04?

**Comment by GenevieveBuckley** [comment]
> cool. I see you've also hard-coded the version string into the toml... that's something we'll need to figure out. thoughts?

I'd like it if we didn't hard code both the version string *and* the list of required packages. I don't know how to do that yet though.

Maybe it's worth a look at [poetry-dynamic-versioning](https://github.com/mtkennerly/poetry-dynamic-versioning) to see how they do it?

**Comment by tlambert03** [comment]
I was actually thinking for the bundle we should pin all of the requirements to specific versions (we don’t want to have to “wonder” which version of a package made it into a bundle), so hard coding that might actually be a good idea.

**Comment by GenevieveBuckley** [comment]
@tlambert03 
On my local Windows, packaging napari does bundle pyrcc properly into the app package & I don't have any problems with the icons either.

Re: `pythonw.exe` vs `python.exe`, I think we can fix this easily by modifying [this line](https://github.com/beeware/briefcase-windows-msi-template/blob/e4d2171d5cc877d6d9e633fe2f18073d21126b4d/%7B%7B%20cookiecutter.formal_name%20%7D%7D/%7B%7B%20cookiecutter.app_name%20%7D%7D.wxs#L65). You can pass in an argument to tell briefcase which template to use, so we could even fork and host a modified template if need be.

**Comment by tlambert03** [comment]
Can you tell me which version of pyside2 you’re bundling and where exactly the pyside2-rcc binary ends up in the bundle?

**Comment by GenevieveBuckley** [comment]
> I was actually thinking for the bundle we should pin all of the requirements to specific versions (we don’t want to have to “wonder” which version of a package made it into a bundle), so hard coding that might actually be a good idea.

I'd like it to just pull that info from `requirements/default.txt` so it is hard coded, but only in one place in the repository.

**Comment by GenevieveBuckley** [comment]
I'm available now for a short video chat if you like, that might be quicker to share info. If that suits I can ping you on zulip?

**Comment by tlambert03** [comment]
We can switch pythonw to python, yes, but then we’ll unfortunately have a terminal window also popping up with napari every time, right?

**Comment by GenevieveBuckley** [comment]
> We can switch pythonw to python, yes, but then we’ll unfortunately have a terminal window also popping up with napari every time, right?

This is exactly what happens with CellProfiler on Windows, for what that's worth

**Comment by GenevieveBuckley** [comment]
There's one other thing you might not have noticed yet @tlambert03 

`briefcase` imports some classes from the standard library `types`. Since napari *also* has a module named `types`, this gets clobbered if you try and run briefcase from an environment where napari has already been pip installed. 

I don't think you've seen this yet because your CI actions pipeline doesn't need to install napari to work. But it would be better for developers to have just a single dev environment while working on napari, where they can run briefcase locally if they want.

Over on my [work in progress branch](https://github.com/GenevieveBuckley/napari/tree/briefcase-packaging) I've just renamed napari `types` to `naparitypes` instead (I'm bad at names, we could always call it something else).

**Comment by tlambert03** [comment]
I _had_ seen this actually and had to rename it too!

**Comment by tlambert03** [comment]
I don't totally get it though... why should that nested module be preferred over the stdlib one?  There are tons of packages that have a `package.types` module...

**Comment by GenevieveBuckley** [comment]
So I have a solution for dynamic version strings in `pyproject.toml` 🎉 

The same solution should also work for keeping the briefcase dependency requirements in synch with what's in requirements.txt

I think we should:
1.  Keep the slightly flexible version requirements for dependencies (eg: `dask[array]>=2.1.0`, not `dask[array]==2.1.0`)
2. Add an output from the github actions that gives us a record of exactly which versions got bundled into the apps

This seems like easily the best of both worlds: we always know exactly which package versions were bundled in the app (because it was reported just after building), and we don't spend time fiddling around working out exactly which combination of requirements will work for each of the different platforms.

**Comment by tlambert03** [comment]
hmm, I'm still not sure.  I think for a bundled app, I'm happy pinning to specific versions, (even if those versions are outdated!). 

having a after-the-fact record of what went in is slightly better than nothing... but it won't stop breakages from occurring.  fwiw, I see app bundling as a very different target than `pip install napari` (which is more like a library that needs to behave in an unknown environment).

**Comment by GenevieveBuckley** [comment]
Ok, we'll leave that idea for now.

We can dynamically write the version string in pyproject.toml like this:
```python
import tomlkit
from napari import __version__

with open('pyproject.toml') as f:
    content = f.read()

doc = tomlkit.parse(content)
doc['tool']['briefcase']['version'] = __version__

with open('pyproject.toml', 'w') as f:
    f.write(tomlkit.dumps(doc))
```

We could add it as a pre-requisite job in the github actions workflow (and put an empty string or obvious placeholder in that section of pyproject.toml)

**Comment by tlambert03** [comment]
updated artifacts: https://github.com/tlambert03/napari/actions/runs/114793446/
should work on mac (with signing caveat mentioned in https://github.com/napari/napari/issues/1074#issuecomment-633303878), and seems to work on windows now too, with working built in console (but visible console window).  I did a kinda weird quick patch method, rather than creating a whole new template... we can definitely change that later if desired, but it works and doesn't require us to maintain our own template. 

Linux is still not working, but have confirmation that it's a ctypes bug in the beeware linux support package, so we'll have to wait on that for a fix.  That said: [I still think using snapcraft](https://github.com/napari/napari/issues/1074#issuecomment-610099849) is the preferable method for linux, so I see the linux briefcase install as the least important.

**Comment by GenevieveBuckley** [comment]
> That said: [I still think using snapcraft](https://github.com/napari/napari/issues/1074#issuecomment-610099849) is the preferable method for linux, so I see the linux briefcase install as the least important.

If we get that working, we might even be able to share that in a useful way here: [Add Snap Packaging #358](https://github.com/beeware/briefcase/issues/358)

**Comment by GenevieveBuckley** [comment]
Re: code-signing & notarization for Mac apps
* See step 4 onwards: https://glyph.twistedmatrix.com/2018/01/shipping-pygame-mac-app.html
* Issue tracking adding notarization to briefcase: https://github.com/beeware/briefcase/issues/365

**Comment by tlambert03** [comment]
you know what, I'm also remembering now something I realized with my previous mac bundling implementation that is almost certainly at play here... namely: once signed, _nothing_ about the bundle can change before the user approves it.  So the resources building themselves and adding a new `_qt_resources.py` file upon app launch is a no-no.  We'll have to build those during CI prior to bundling.  I think _that's_ actually why you need to use that terminal workaround I mentioned in a previous post (but also an easy fix).

**Comment by tlambert03** [comment]
cc @goanpeca  ... this is the thread that had most of our original discussion about how to bundle an environment that was extendable.  here is the https://github.com/napari/napari/issues/1074#issuecomment-606162357 where we briefly toyed with the [conda-bundler](https://github.com/tlambert03/conda-bundler) idea (not sure how similar that is to spyder approaches though)...
